### PR TITLE
Need MAP_NORESERVE in VirtualReserve on illumos

### PR DESCRIPTION
**Pull request by gwr**
Need MAP_NORESERVE in VirtualReserve on illumos (and maybe some other platforms)

Earlier I thought the issue was that `GetAvailablePhysicalMemory()` was returning a value that's too large.
However, after looking at how this works on Linux, I found that on Linux the mmap call for a 256GB memory range actually succeeds. That was with testing on an 8GB VM, so clearly the call on Linux is NOT allocating physical pages for that range.  To get that behavior on illumos, one should pass MAP_NORESERVE.

I might suggest changing the issue title and commit first line, but I don't appear to be able to do that.
I'd prefer: Need MAP_NORESERVE in VirtualReserve for illumos
or something like that.

### CODE REVIEW

File: src/coreclr/gc/unix/gcenv.unix.cpp [hunk 1]
```diff
@@ -580,7 +580,13 @@ static void* VirtualReserveInner(size_t size, size_t alignment, uint32_t flags,
     }
 
     size_t alignedSize = size + (alignment - OS_PAGE_SIZE);
-    void * pRetVal = mmap(nullptr, alignedSize, PROT_NONE, MAP_ANON | MAP_PRIVATE | hugePagesFlag, -1, 0);
+    int mapflags = MAP_ANON | MAP_PRIVATE | hugePagesFlag;
+#if defined(MAP_NORESERVE) && defined(__sun)
```
Comment (am11) [inline src/coreclr/gc/unix/gcenv.unix.cpp:584]: nit: to be consistent with TARGET_LINUX and TARGET_APPLE in nearby context.
```suggestion
#if defined(MAP_NORESERVE) && defined(TARGET_SUNOS)
```

Comment (gwr) [reply src/coreclr/gc/unix/gcenv.unix.cpp:584]: OK, though I see a lot of cases where the code currently uses `__sun` as the pre-processor symbol to check.
If we're going to use TARGET_SUNOS instead, we should probably change them all.

Personally, I'd skip it for now, and maybe later add a TARGET_ILLUMOS (less confusing name)
and then go change all the __sun checks.

**Review by am11 — APPROVED** [review]
One style nit

**Comment by gwr** [comment]
Is #104211 the better issue to post this against?  Or list both issues?

**Comment by am11** [comment]
Yes, it is related. We are eliminating the need of hard heap limit in GC regions calculation on illumos. Both Physical and Virtual mems participate in this calculation, so discussion is related.
BTW, it is not a requirement to have an issue, just good to have thing to help future reader if they run into a similar situation.

File: src/coreclr/gc/unix/gcenv.unix.cpp (hunk 1 above)
Comment (am11) [reply src/coreclr/gc/unix/gcenv.unix.cpp:584]: np, it's ok to keep it as-is.

* we apply "what's used in the nearby context" logic for consistency. So when there is  {HOST,TARGET}_XX used in a file, those are preferred.
* TARGET_SUNOS (more permissive) is better than TARGET_ILLUMOS unless we can determine solaris implementation is going to differ.

Comment (gwr) [reply src/coreclr/gc/unix/gcenv.unix.cpp:584]: Let me try the suggest on my setup to make sure it works OK.

Comment (gwr) [reply src/coreclr/gc/unix/gcenv.unix.cpp]: Yup, that works fine.  Push it! :)

**Comment by am11** [comment]
linux: https://man7.org/linux/man-pages/man2/mmap.2.html

```
MAP_NORESERVE
              Do not reserve swap space for this mapping.  When swap
              space is reserved, one has the guarantee that it is
              possible to modify the mapping.  When swap space is not
              reserved one might get SIGSEGV upon a write if no physical
              memory is available.  See also the discussion of the file
              /proc/sys/vm/overcommit_memory in [proc(5)](https://man7.org/linux/man-pages/man5/proc.5.html).  Before Linux
              2.6, this flag had effect only for private writable
              mappings.
```

illumos: https://illumos.org/man/2/mmap

```
       The MAP_NORESERVE option specifies that no swap space be reserved for a
       mapping. Without this flag, the creation of a writable MAP_PRIVATE
       mapping reserves swap space equal to the size of the mapping; when the
       mapping is written into, the reserved space is employed to hold private
       copies of the data. A write into a MAP_NORESERVE mapping produces results
       which depend on the current availability of swap space in the system.  If
       space is available, the write succeeds and a private copy of the written
       page is created; if space is not available, the write fails and a SIGBUS
       or SIGSEGV signal is delivered to the writing process.  MAP_NORESERVE
       mappings are inherited across fork(); at the time of the fork(), swap
       space is reserved in the child for all private pages that currently exist
       in the parent; thereafter the child's mapping behaves as described above.
```

Please make sure this change is what the GC regions code is expecting and will behave as intended.

A safer solution is to disable `USE_REGIONS` for illumos as we have for macOS. For now, until we get the libraries ported and run tests on the platform which exercise GC more thoroughly.

**Comment by gwr** [comment]
> Please make sure this change is what the GC regions code is expecting and will behave as intended.

I am testing with this change and things look good so far.
(I still could use help on next steps for testing and porting.)

> A safer solution is to disable `USE_REGIONS` for illumos as we have for macOS; for now (until we get the libraries ported and run tests on the platform which exercise GC more thoroughly).

The Linux VM I'm using for comparison has 8GB of RAM and is configured with about 1GB of swap space.
Given that, I don't see how Linux could actually be reserving the pages when dotnet mmap 256G of space.

I believe passing `MAP_NORESERVE` brings the illumos behavior as close as possible to what Linux does here.

**Comment by am11** [comment]
> I am testing with this change and things look good so far.

Hello world is only a measure of telling "runtime starts up and gracefully shuts down". It is not telling us whether GC regions feature is working properly. There are plenty of scenarios which will exercise the GC more holistically once we will run the tests.

**Comment by am11** [comment]
We can do it in the header where macOS is excluded https://github.com/dotnet/runtime/issues/34944#issuecomment-2200650644 (cmake only has defined for standalone GC, which we can skip for now as that is not the primary one which links to runtime).

**Comment by gwr** [comment]
> We can do it in the header where macOS is excluded [#34944 (comment)](https://github.com/dotnet/runtime/issues/34944#issuecomment-2200650644) (cmake only has defined for standalone GC, which we can skip for now as that is not the primary one which links to runtime).

OK, that worked. With that change, the largest `mmap` VirtualReserve size was 512MB, which works OK without `MAP_NORESERVE`.  We can revisit this if and when we may decide to enable `USE_REGIONS` later.

**Comment by gwr** [comment]
We may want this later.  Put it in "draft" state or something?  otherwise just close.

**Pull request closed without merging**